src/renderer.rs

```diff
@@ -644,31 +655,30 @@ pub fn ray_intersect(
         Wrapping::ClampToEdge,
         Wrapping::ClampToEdge,
     );
-    let depth_material = DepthMaterial {
-        render_states: RenderStates {
-            write_mask: WriteMask {
-                red: true,
-                ..WriteMask::DEPTH
-            },
-            ..Default::default()
-        },
+    let mut material = IntersectionMaterial {
         ..Default::default()
     };
-    let depth = RenderTarget::new(
+    let result = RenderTarget::new(
         texture.as_color_target(None),
         depth_texture.as_depth_target(),
     )
     .clear(ClearState::color_and_depth(1.0, 1.0, 1.0, 1.0, 1.0))
     .write::<RendererError>(|| {
-        for geometry in geometries {
-            render_with_material(context, &camera, &geometry, &depth_material, &[]);
+        for (id, geometry) in geometries.into_iter().enumerate() {
+            material.geometry_id = id as i32;
+            render_with_material(context, &camera, &geometry, &material, &[]);
         }
         Ok(())
     })
     .unwrap()
-    .read_color::<[f32; 4]>()[0][0];
+    .read_color::<[f32; 4]>()[0];
+    let depth = result[0];
     if depth < 1.0 {
-        Some(position + direction * depth * max_depth)
+        Some(IntersectionResult {
```

Comment:
The IDs here are limited by the maximum safe integer a f32 can hold (approximately 24 bits). However, use cases where the IDs ever get that high are probably extremely rare, so it probably isn't worth the additional complexity that fixing that issue would incur (what's currently done in #479). Maybe this limitation should be added to the documentation somewhere?

Reply:
Yeah, I was thinking 24 bit integer would be enough (16,777,215 instances is probably rare 🙂), but if we also have to output primitive ids, it's probably better to support the whole u32/i32 range. I fixed that in [439261b](https://github.com/asny/three-d/pull/494/commits/439261b3f75d41a6289fee1ec23c049e46313c96)